[Timothy Jonathan Koei] iP

t1mzzz opened the pull request:
## **Echo**

***Echo is a Chatbot that is able to store and keep track of tasks inputted by the user.***

> "Time is what we want most, but what we use worst." - William Penn ([source](https://www.timecamp.com/blog/2022/01/quotes-about-time/))

Struggling to use time well? Tired of keeping track of tasks? Fret not, Echo is here to help you monitor your tasks and improve your time management! AND, it's

- text-based
- ~~EASY~~ ***SUPER EASY*** to use
- ***SUPER FAST*** to learn
- ***FREE*** 😮 

To use Echo, here is a simple guide:

1. Download it from [here](https://github.com/t1mzzz/ip/releases/download/A-Jar/duke.jar).
2. Open the file.
3. Store your tasks.
4. Start managing your tasks conveniently! 👍

Duke's features include:
- [x] Managing Tasks
- [x] GUI

____________________________________________________________________________________________________________________________________
Provided below is the `main` method of Duke:
```java
public static void main(String[] args) {
    new Duke(System.getProperty("user.dir") + "/data/tasks.txt").run();
}
```

Commit:
Add Gradle support

t1mzzz committed:
Level-1

t1mzzz committed:
deleted static in functions

t1mzzz committed:
Level-2

t1mzzz committed:
Level-3

t1mzzz committed:
Level-3

t1mzzz committed:
Level-4

t1mzzz committed:
A-TextUiTesting

t1mzzz committed:
formatting and halfway through Level-5

t1mzzz committed:
Level-5

t1mzzz committed:
Level-6

t1mzzz committed:
Level-6

t1mzzz committed:
A-TextUiTesting

t1mzzz committed:
Fix formatting

t1mzzz committed:
Add load and save

Tasks are not loaded and stored upon opening and closure of the
application.

Loading and saving of tasks allow users to store their past tasks and
retain it.

Let's,
* load tasks from file as application starts
* save tasks in file after application closes

t1mzzz committed:
Add dates

Dates and times are stored as strings in Duke.

Consistent formatting and simpler usage is achieved through the use of
LocalDate.

Let's,
* use LocalDate consistently throughout the application
* switch from YYYY-MM-DD to a tidier format when printed MMM-DD-YYYY

t1mzzz committed:
Use More OOP (A-MoreOOP)

t1mzzz committed:
Organize into Packages (A-Packages) and Fix Storage

t1mzzz committed:
Fix README.md

t1mzzz committed:
Merge branch 'add-gradle-support'

t1mzzz committed:
Use Gradle (A-Gradle)

t1mzzz committed:
Add JUnit Tests (A-JUnit)

t1mzzz committed:
Fix formatting

t1mzzz committed:
Add JavaDoc comments (A-JavaDoc)

t1mzzz committed:
Tweak the code to comply with a coding standard (A-CodingStandard)

t1mzzz committed:
Add Find (Level-9)

t1mzzz committed:
Reset tasks.txt

t1mzzz committed:
Merge branch 'branch-Level-9'

t1mzzz committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Task.java

t1mzzz committed:
Parser class: add import FindCommand

t1mzzz committed:
Fix formatting

t1mzzz committed:
Update Chatbot name and fix format

Echomo-Xinyu commented on README.md:
````diff
@@ -13,7 +13,7 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
    In the same dialog, set the **Project language level** field to the `SDK default` option.
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/duke.Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
    ```
````
Very good that you remember to edit the README after altering the original structure. But why not `src/main/java/duke/Duke.java` though?

Echomo-Xinyu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (Exception e) {
```
I think it may not be a good practice to catch `Exception` as a "pokemon catch". Would you want to consider specifying the type of exception you want to handle?

Echomo-Xinyu commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,108 @@
+                }
+            }
+        } else {
+            File parent = new File(DIR + "/data");
```
Would you want to use a more specific variable name such as `parentDir` to make it more suited to the context?

Echomo-Xinyu replied:
Would you want to use a more specific variable name such as `parentDir` to make it more suited to the context?

Echomo-Xinyu commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,55 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        String[] split = cmd.split(" ", 2);
+        String type = split[0];
+        String desc = split[1];
```
Maybe you can consider changing `desc` to another variable name to make its meaning clearer?

Echomo-Xinyu replied:
Maybe you can consider changing `desc` to another variable name to make its meaning clearer?

Echomo-Xinyu commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,47 @@
+            ui.showMessage("  " + curr);
+        } else {
+            ui.showMessage("This task has not been done in the first place.");
+            ui.showMessage("  " + curr);
```
I have noticed several places with such `ui.showMessage("String")`. Since the `Ui` class should be handling all the interactions with the users, I think it would be better if you can abstract this section out and have a corresponding function in `Ui` to handle this request? But since it is about the design part, it is really up to you to decide how you wish to organize your code. ;)

Echomo-Xinyu commented on README.md:
```diff
@@ -13,7 +13,7 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
 1. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
    In the same dialog, set the **Project language level** field to the `SDK default` option.
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/duke.Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```
Minor point but I believe the correct new path should be `src/main/java/duke/Duke.java`?

Jnwkm replied:
agreed with the above comment.

t1mzzz replied:
Thanks for the reminder!

Echomo-Xinyu reviewed (commented):
In general, this is a very well-written piece of work and I can really see your efforts in designing and implementing all the relevant features. Besides several minor issues as commented, I think this code is of great quality. Good job! 👍

weiquany commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,93 @@
+                + "     |  ____|| |     |  __  || |  | |\n"
+                + "     | |____ | |____ | |  | || |__| |\n"
+                + "     |______||______||_|  |_||______|\n";
+        content = logo + "\n" + INDENT + "Hello! I'm Echo\n" + INDENT + "What can I do for you?\n";
```
Will it be better to use `System.lineSeparator()` rather than `\n`? I am unsure if `\n` can work on different OSes.

weiquany commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,82 @@
+                default:
+                    return new AddCommand(fullCommand);
+                }
+            } catch (Exception e) {
```
I don't think this violates the CS2103 code quality standards and I quite like this way of handling errors in this particular case, but will it be better to catch some of the possible exceptions before the "catch-all" statement?

weiquany reviewed (commented):
Great job overall on adhering to the CS2103 code quality standards! I had a difficult time looking for possible mistakes, and may have made mistakes in my review, do feel free to correct me!

t1mzzz committed:
Complete JavaFX part 1

t1mzzz committed:
Complete JavaFX tutorial part 2, part 3 (iteration 1 & 2)

t1mzzz committed:
Complete JavaFX tutorial part 3

t1mzzz committed:
Add GUI

t1mzzz committed:
Fix format

t1mzzz committed:
Show welcome
Delete unused variables and methods

t1mzzz committed:
Change user and bot pictures

t1mzzz committed:
Add close window

Jnwkm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class Duke {
+
+    private static final String FILE_PATH = System.getProperty("user.dir") + "/data/tasks.txt";
```
Is this really necessary if you use relative paths? Not to say it's wrong but maybe could be simpler :)

Jnwkm reviewed (commented):
Overall, it's very good! Impressed with the implementation of all the extra optional classes and you've kept it very clean!

t1mzzz committed:
Resolve issues from PR feedback

emptygx commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,73 @@
+     * @param i The index of the task.
+     * @return The task corresponding to the index.
+     */
+    public Task get(int i) {
```
Might be better to be more specific on the function name given there are other getter functions in this class.
```suggestion
    public Task getTask(int i) {
```

emptygx commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,44 @@
+
+    private Scanner sc = new Scanner(System.in);
+
+    /**
+     * Shows the message when a task is added.
+     *
+     * @param size The size of the TaskList.
+     * @param desc The description of the task.
+     */
```
Missing a return section for the string in the JavaDoc.

emptygx commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,44 @@
+        return output;
+    }
+
+    /**
+     * Shows an error message with an indentation.
+     *
+     * @param errMessage The error message to be printed.
+     */
```
Missing a return section for the string in the JavaDoc.

emptygx commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,44 @@
+        return errMessage;
+    }
+
+    /**
+     * Shows the closing message when user finishes using the program.
+     */
```
Missing a return section for the string in the JavaDoc.

emptygx commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        if (!details.contains("/by")) {
+            throw new DukeException("OOPS!!! The deadline is required. (/by)");
+        }
+        String[] split = details.split(" /");
+        String desc = split[0];
+        if (desc.equals("") || desc.equals(" ")) {
+            throw new DukeException("OOPS!!! Description of deadline is required.");
+        }
+        String by = split[1].split(" ", 2)[1];
+        Deadline deadline = new Deadline(desc, by);
+        tasks.add(deadline);
+        return deadline.toString();
```
Can consider parsing the deadline date within the Parser class since that is the role of the class, but it is a design decision so you can decide! 😄

emptygx reviewed (commented):
LGTM overall! Just as mentioned by others, the project has clean code with little coding standard and quality violations. Just some minor nits and changes to consider, but good job altogether! 👍

t1mzzz committed:
Generate new JAR file

t1mzzz committed:
Use Assertions

Some methods require assertions as there may be invalid parameters in
them.

Using assertions allow for checking where various assumptions in the
code are valid or not.

Let's utilize the power of assertions to document important assumptions
that should hold at various points in the code.

t1mzzz committed:
Use assertions

Some methods require assertions as there may be invalid parameters in
them.

Using assertions allow for checking where various assumptions in the
code are valid or not.

Let's utilize the power of assertions to document important assumptions
that should hold at various points in the code.

t1mzzz committed:
Merge branch 'branch-A-Assertions' of https://github.com/t1mzzz/ip into branch-A-Assertions

t1mzzz committed:
Merge pull request #1 from t1mzzz/branch-A-Assertions

Use Assertions

Some methods require assertions as there may be invalid parameters in
them.

Using assertions allows for checking where various assumptions in the
code is valid or not.

Let's utilize the power of assertions to document important assumptions
that should hold at various points in the code.

t1mzzz committed:
Improve code quality

Some methods contain magic numbers and magic literals. These have to be
fixed.

Using static final variables allow for improvement in code quality.

Let's critically examine the code and refactor to improve the
code quality where necessary.

t1mzzz committed:
Merge pull request #2 from t1mzzz/branch-A-CodeQuality

Improve code quality

Some methods contain magic numbers and magic literals. These have to be
fixed. 

Using static final variables allow for improvement in code quality.

Let's critically examine the code and refactor to improve the 
code quality where necessary.

t1mzzz committed:
Set up CI

t1mzzz committed:
Add sort based on date

t1mzzz committed:
Fix Coding Standard and JavaDoc

t1mzzz committed:
Fix CI

t1mzzz committed:
Fix JavaDoc

t1mzzz committed:
Handle more errors

t1mzzz committed:
Add Ui.png to docs directory

t1mzzz committed:
Add a User Guide

t1mzzz committed:
Fix User Guide and Ui example image

t1mzzz committed:
Fix spacing of example image

t1mzzz committed:
Fix formatting in User Guide

t1mzzz committed:
Fix User Guide

t1mzzz committed:
Disable resize

t1mzzz committed:
Fix adding event and deadline bug

t1mzzz committed:
Fix Checkstyle issue
